Issue #6106 - Fix WebSocket/CDI integration in Jetty 10

lachlan-roberts opened the pull request:
Issue #6106

lachlan-roberts committed:
Issue #6106 - add tests for WebSocket/CDI integration.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Issue #6106 - fix WebSocket/CDI integration

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

joakime reviewed (changes requested):
The `javax.websocket` components need a `DecoratedObjectFactory.decorate()` call as well, if they were created by Jetty. (as opposed to being handed to Jetty already instantiated by the application).

* javax.websocket.Encoder
* javax.websocket.Decoder
* javax.websocket.ClientEndpointConfig.Configurator
* javax.websocket.server.ServerEndpointConfig.Configurator

joakime commented:
We should also support CDI on `Extension` instance we create as well.

joakime commented:
There's also (in jetty-10+) the `org.eclipse.jetty.websocket.core.server.Handshaker.newInstance()` that should go through the DecoratedObjectFactory.

gregw reviewed (changes requested):
tell me when ready to review again...

lachlan-roberts committed:
Issue #6106 - Decorate javax.websocket Encoder & Decoders.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Issue #6106 - always decorate javax.websocket Configurators

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Issue #6106 - add testing for CDI with javax.websocket Encoder & Decoder

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Fix broken websocket tests.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts commented:
@joakime we already use the `DecoratedObjectFactory` to create extensions. And I don't think we need it for `Handshaker.newInstance()` because that is just creating an internal class, you can't provide a custom handshaker for that method to create.

joakime reviewed (changes requested):
I'd like to see a testcase of a `javax.websocket` Server Endpoint that has an `@Inject` of the servlet `HttpSession` as that is a very common use case.

gregw reviewed (dismissed):
wow! Big!!!

lachlan-roberts committed:
Issue #6106 - Decorate javax.websocket components during the websocket upgrade.

Decoration should not be delayed until onOpen it should be during the call to
the WebSocketUpgradeFilter or WebSocketServlet.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts commented:
> I'd like to see a testcase of a javax.websocket Server Endpoint that has an @Inject of the servlet HttpSession as that is a very common use case.

@joakime I think this isn't going to be very simple, so should be taken out to a separate issue.

We would need to somehow propagate the `@RequestScope` context for the lifecycle of the websocket connection. Otherwise the websocket will try access the injected `HttpSession` from a different thread from the `WebSocketUpgradeFilter` so it won't be in CDI scope anymore and we get an error trying to use it. We decorate the endpoint instance during the call to `WebSocketUpgradeFilter`, but it only injects a weld CDI wrapper of the object which doesn't work when used in a different thread.

joakime commented:
> > I'd like to see a testcase of a javax.websocket Server Endpoint that has an @Inject of the servlet HttpSession as that is a very common use case.
> 
> @joakime I think this isn't going to be very simple, so should be taken out to a separate issue.
> 
> We would need to somehow propagate the `@RequestScope` context for the lifecycle of the websocket connection. Otherwise the websocket will try access the injected `HttpSession` from a different thread from the `WebSocketUpgradeFilter` so it won't be in CDI scope anymore and we get an error trying to use it. We decorate the endpoint instance during the call to `WebSocketUpgradeFilter`, but it only injects a weld CDI wrapper of the object which doesn't work when used in a different thread.

That's not how it works.
The `@Inject HttpSession httpSession` will cause the `HttpSession` from the servlet layer to be added/set to the Endpoint during our `DecoratedObjectFactory.decorate(endpoint)` call.

The (server) endpoint we create should have the HttpSession populated from the servlet layer during the time of the upgrade itself.
There is no update to that endpoint by CDI after the upgrade is complete.

lachlan-roberts commented:
@joakime it is how it works, I have a test for this which fails (which I have not committed to this branch).

It successfully injects the `HttpSession` into the WebSocket Endpoint, but its class is `org.jboss.weldx.servlet.http.HttpSession$$Proxy$_$$_WeldClientProxy`. It works fine if accessed in the same thread that's doing the call to `doFilter()`, but if accessed from another thread, it will throw `ContextNotActiveException`.

```
Caused by: org.jboss.weld.contexts.ContextNotActiveException: WELD-001303: No active contexts for scope type javax.enterprise.context.SessionScoped
	at org.jboss.weld.manager.BeanManagerImpl.getContext(BeanManagerImpl.java:647)
	at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.getIfExists(ContextualInstanceStrategy.java:89)
	at org.jboss.weld.bean.ContextualInstance.getIfExists(ContextualInstance.java:63)
	at org.jboss.weld.bean.proxy.ContextBeanInstance.getInstance(ContextBeanInstance.java:87)
	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:105)
	at org.jboss.weldx.servlet.http.HttpSession$$Proxy$_$$_WeldClientProxy.getId(Unknown Source)
	at org.eclipse.jetty.cdi.tests.websocket.JavaxWebSocketCdiTest$CdiHttpSessionSocket.onMessage(JavaxWebSocketCdiTest.java:401)
```

joakime commented:
> @joakime it is how it works, I have a test for this which fails (which I have not committed to this branch).
> 
> It successfully injects the `HttpSession` into the WebSocket Endpoint, but its class is `org.jboss.weldx.servlet.http.HttpSession$$Proxy$_$$_WeldClientProxy`. It works fine if accessed in the same thread that's doing the call to `doFilter()`, but if accessed from another thread, it will throw `ContextNotActiveException`.
> 
> ```
> Caused by: org.jboss.weld.contexts.ContextNotActiveException: WELD-001303: No active contexts for scope type javax.enterprise.context.SessionScoped
> 	at org.jboss.weld.manager.BeanManagerImpl.getContext(BeanManagerImpl.java:647)
> 	at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.getIfExists(ContextualInstanceStrategy.java:89)
> 	at org.jboss.weld.bean.ContextualInstance.getIfExists(ContextualInstance.java:63)
> 	at org.jboss.weld.bean.proxy.ContextBeanInstance.getInstance(ContextBeanInstance.java:87)
> 	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:105)
> 	at org.jboss.weldx.servlet.http.HttpSession$$Proxy$_$$_WeldClientProxy.getId(Unknown Source)
> 	at org.eclipse.jetty.cdi.tests.websocket.JavaxWebSocketCdiTest$CdiHttpSessionSocket.onMessage(JavaxWebSocketCdiTest.java:401)
> ```

OK, You've sold me on this.

We should ...
1. Open a new issue for the inject of HttpSession in endpoints.
2. This new HttpSession testcase should be disabled and labelled to look at the new issue number.

lachlan-roberts committed:
Disable test for WebSocket CDI HttpSession injection and link to issue.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts commented:
> 1. Open a new issue for the inject of HttpSession in endpoints.
> 2. This new HttpSession testcase should be disabled and labelled to look at the new issue number.

@joakime I have made these changes, see #6174 for the new issue.

joakime commented on jetty-websocket/websocket-javax-server/src/main/java/org/eclipse/jetty/websocket/javax/server/internal/JavaxWebSocketCreator.java:
```diff
@@ -159,8 +159,6 @@ public Object createWebSocket(ServerUpgradeRequest req, ServerUpgradeResponse re
             Class<?> endpointClass = config.getEndpointClass();
             Object endpoint = config.getConfigurator().getEndpointInstance(endpointClass);
-            // Do not decorate here (let the Connection and Session start first)
-            // This will allow CDI to see Session for injection into Endpoint classes.
```
Why was this comment removed?
Seems like an important reminder for future devs.

lachlan-roberts replied:
The comment is incorrect. We don't support the injection of the WebSocket Session into Endpoint classes.

We don't decorate there because we decorate in the `FrameHandlerFactory` instead, and we do this because it can be done in the same way between the Jetty and Javax sides because you can provide your own creator on the Jetty side so we cannot decorate there.

joakime reviewed (commented):
That comment removal seems odd.

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-6106-WebSocketCDI
